pop_tools/fill.py

```diff
@@ -183,3 +223,93 @@ def _iterative_fill_POP_core(nlat, nlon, var, fillmask, missing_value, tol, ltri
+    for j in range(0, nlat):
+        zoncnt[j] = np.sum(np.where(fillmask[j, :], 0, 1))
+        zonavg[j] = np.sum(np.where(fillmask[j, :], 0, var[j, :]))
+        if zoncnt[j] != 0:
+            zonavg[j] = zonavg[j] / zoncnt[j]
+
+    # Fill missing zonal averages for rows that are entirely land
+    for j in range(0, nlat - 1):  # northward pass
+        if zoncnt[j] > 0 and zoncnt[j + 1] == 0:
+            zoncnt[j + 1] = 1
+            zonavg[j + 1] = zonavg[j]
+    for j in range(0, nlat - 1):  # southward pass
+        jrev = nlat - 1 - j
+        if zoncnt[jrev] > 0 and zoncnt[jrev - 1] == 0:
+            zoncnt[jrev - 1] = 1
+            zonavg[jrev - 1] = zonavg[jrev]
+
+    # Replace the input array missing values with zonal average as first guess
+    for j in range(0, nlat):
+        for i in range(0, nlon):
+            if fillmask[j, i]:
+                var[j, i] = zonavg[j]
+
+    # Now do the iterative 2D fill
+    res = np.zeros((nlat, nlon))  # work array hold residuals
+    res_max = tol
+    iter = 0
+    while iter < max_iter and res_max >= tol:
+        res = res * 0.0  # reset the residual to zero for this iteration
+
+        # assume bottom row is all land, leave it set to zonal average
+        # deal with top row separately below
+        for j in range(1, nlat - 1):
+            jm1 = j - 1
+            jp1 = j + 1
+
+            for i in range(0, nlon):
+                if fillmask[j, i]:
+                    im1 = i - 1
+                    if i == 0:  # assume periodic in x
+                        im1 = nlon - 1
+                    ip1 = i + 1
+                    if i == nlon - 1:
+                        ip1 = 0
+
+                    # this is SOR
+                    res[j, i] = (
+                        var[j, ip1] + var[j, im1] + var[jm1, i] + var[jp1, i] - 4.0 * var[j, i]
+                    )
+                    var[j, i] = var[j, i] + rc * 0.25 * res[j, i]
+
+        # do the top row if there was some valid data there in the input
+        # otherwise leave it set to zonal average of northernmost row with valid data
+        if zoncnt[nlat - 1] > 1:
+            j = nlat - 1
+            jm1 = j - 1
+            jp1 = j
+            for i in range(0, nlon):
+                if fillmask[j, i]:
+                    im1 = i - 1
+                    if i == 0:
+                        im1 = nlon - 1
+                    ip1 = i + 1
+                    if i == nlon - 1:
+                        ip1 = 0
+                    io = nlon - 1 - i
+
+                    if ltripole:  # use cross-pole periodicity
+                        res[j, i] = (
+                            var[j, ip1] + var[j, im1] + var[jp1, io] + var[jm1, i] - 4.0 * var[j, i]
+                        )
+                        var[j, i] = var[j, i] + rc * 0.25 * res[j, i]
+                    else:  # do a 1D smooth on pole row
+                        res[j, i] = var[j, ip1] + var[j, im1] - 2.0 * var[j, i]
```

Comment:
Can you assume that the top row is land for all levels, i.e., `KMT==0`, if `ltripole == False`? The code is making a similar assumption for the bottom row, independent of `ltripole`. I just verified that `KMT==0` in the top row of our gx1v[67] and gx3v7 grids. If you make this assumption then the code can be simplified. The `ltripole` conditional can be brought outside the `i` loop, and the `else` clause can be dropped.

I do think you would still need to check `zoncnt[nlat - 1]` for `ltripole == True`, because deep enough layers might be all land along the tripole seam.

Reply:
My use case is filling WOA land on its own grid, so the North pole may have land and ocean points.